### ISSUE TITLE
[Discussion] Bootstrapping from mrustc

### ISSUE THREAD

**Issue by dongcarl** [rust-bitcoin/rust-bitcoin#197]
I've been trying my luck at bootstrapping rust 1.20.0 from [mrustc](https://github.com/thepowersgang/mrustc). I wanted to open this thread so we can discuss our findings and methodologies.

Here are the resources I've found, where others are attempting the same thing:

### Slackware (0.19.0 - 1.30.1)
https://github.com/aclemons/slack-rust-bootstrap

### GUIX (plan only)
https://bootstrapping.miraheze.org/wiki/Boostrapping_Specific_Languages#Rust

### Arch Linux (0.19.0 only)
`mrustc` and `minicargo`: https://aur.archlinux.org/packages/mrustc-git
`rust=1.19.0`: https://aur.archlinux.org/packages/rust-mrustc/

From my experience, the build instructions are quite fragile, and would perhaps be improved when https://github.com/thepowersgang/mrustc/pull/88 gets merged.

**Comment by dongcarl** [comment]
`mrustc` seems to want very specific versions of toolchains and dependencies.

Only gcc versions < 8 works.
OpenSSL 1.0.1, 1.0.2, and 1.1.0, or LibreSSL 2.5.0, 2.5.1, 2.5.2, 2.5.3, and 2.5.4 needs to be the version that pkg-config chooses by default.

**Comment by dongcarl** [comment]
Learning more...

Best way to do OpenSSL is to compile it yourself with the instructions [here](https://github.com/sfackler/rust-openssl/tree/v0.9.12#linux), then set `OPENSSL_DIR` to the prefix.

Also, the Makefile is not very robust... Not everything can be parallel, and some builds with break if you do `-j*`, build without `-j` if you're unsure.

**Comment by dongcarl** [comment]
I have successfully produced a 1.20.0 rustc bootstrapped from mrustc: https://we.tl/t-dPn25IcgIL

**Comment by sgeisler** [comment]
Do you have something like recorded steps/bash_history that you can share? I think that's much more interesting than the actual build.

**Comment by dongcarl** [comment]
Of course, I'm in the process of determining which of my many overlapping hacks and patches were actually needed. Hopefully it's somewhat simple in the end.

**Comment by elichai** [comment]
FYI. mrustc now supports 1.29
Although I have some problems with it with latest GCC: https://github.com/thepowersgang/mrustc/issues/123

**Comment by apoelstra** [comment]
Relevant: https://github.com/Rust-GCC/gccrs

**Comment by devrandom** [comment]
mrustc now supports 1.39
